### PR TITLE
[Triton] Add verifier for triton call/dispatch ops

### PR DESCRIPTION
**Pull request by ezhulenev**
* Also add explicit HAL pipeline layout and automatically infer it from the Triton signature
* Automatically update Triton function signature to ABI-compatible type

This is part of the work to build end-to-end Triton pipeline hooked to `triton.executable` operation.

### CODE REVIEW

File: compiler/src/openxla/compiler/nvgpu/Dialect/TritonFlow/Transforms/OutlineTritonCalls.cpp [hunk 1]
```diff
@@ -28,12 +38,135 @@ using namespace mlir;
...
+  };
+
+  SmallVector<IndexedArg> args;
+  for (auto indexed : llvm::enumerate(callee.getArgumentTypes()))
+    args.push_back({indexed.index(), indexed.value()});
+
+  // Partition arguments types, so that all scalars pushed to the end.
+  auto firstScalarArg = std::stable_partition(
+      args.begin(), args.end(),
+      [](IndexedArg indexed) { return !indexed.type.isIntOrIndex(); });
```
Comment (chsigg) [inline compiler/src/openxla/compiler/nvgpu/Dialect/TritonFlow/Transforms/OutlineTritonCalls.cpp:98]: Would it make sense to pull `type.getIntOrIndex()` into a `isScalarArg()` function (I've seen one other occurrence above)?

**Review by chsigg — APPROVED** [review]
Just some minor nits...

File: compiler/src/openxla/compiler/nvgpu/Dialect/TritonFlow/IR/TritonFlowOps.cpp [hunk 2]
```diff
@@ -42,6 +48,140 @@ static LogicalResult verifyOpDynamicDims(Operation *op, ValueRange values,
   return success();
 }
 
+static LogicalResult verifyArgumentTypes(Operation *op, TypeRange tritonArgs,
+                                         int64_t tritonOffset, TypeRange args,
+                                         int64_t offset = 0) {
+  assert(tritonArgs.size() == args.size() && "type ranges must have same size");
+  for (auto pair : llvm::enumerate(llvm::zip(tritonArgs, args))) {
```
[the remark this answers is not in the document]
Comment (ezhulenev) [reply compiler/src/openxla/compiler/nvgpu/Dialect/TritonFlow/IR/TritonFlowOps.cpp]: Done

File: compiler/src/openxla/compiler/nvgpu/Dialect/TritonFlow/IR/TritonFlowOps.cpp [hunk 3]
```diff
@@ -158,9 +313,25 @@ LogicalResult DispatchOp::verify() {
...
-  if (!symbolTable.lookupNearestSymbolFrom(*this, getCalleeAttr()))
+  Operation *op = getOperation();
+
+  auto tritonFunc =
+      symbolTable.lookupNearestSymbolFrom<triton::FuncOp>(op, getCalleeAttr());
+  if (!tritonFunc)
     return emitOpError() << "refers to an unknown Triton function: "
                          << getCalleeAttr();
+
+  // TODO(ezhulenev): Skip verifier until we didn't migrate to an end-to-end
```
[the remark this answers is not in the document]
Comment (ezhulenev) [reply compiler/src/openxla/compiler/nvgpu/Dialect/TritonFlow/IR/TritonFlowOps.cpp]: Added a shorter TODO for myself, will clean it in the next few PRs.

File: compiler/src/openxla/compiler/nvgpu/Dialect/TritonFlow/Transforms/OutlineTritonCalls.cpp (hunk 1 above)
Comment (ezhulenev) [reply compiler/src/openxla/compiler/nvgpu/Dialect/TritonFlow/Transforms/OutlineTritonCalls.cpp]: Done

**Pull request merged**